proxy/src/control/cache.rs

```diff
@@ -37,54 +46,85 @@ impl<T> Exists<T> {
-            where T: Copy + Eq + std::hash::Hash, I: Iterator<Item = T>, F: FnMut(T, CacheChange)
+        fn update_inner<K, V, I, F>(
+            values: &mut IndexMap<K, V>,
+            iter: I,
+            on_change: &mut F
+        )
+        where
+            K: Eq + Hash + Copy + Clone,
+            V: PartialEq + Clone,
+            I: Iterator<Item = (K, V)>,
+            F: FnMut((K, V), CacheChange),
         {
-            for value in iter {
-                if values.insert(value) {
-                    on_change(value, CacheChange::Insertion);
+            for (key, value) in iter {
+                match values.insert(key, value.clone()) {
+                    // If the returned value is equal to the inserted value,
+                    // then the inserted key was already present in the map
+                    // and the value is unchanged. Do nothing.
+                    Some(ref old_value) if old_value == &value => {},
+                    // If `insert` returns `Some` with a different value than
+                    // the one we inserted, then we changed the value for that
+                    // key.
+                    Some(_) =>
+                        on_change((key, value), CacheChange::Modification),
+                    // If `insert` returns `None`, then there was no old value
+                    // previously present. Therefore, we inserted a new value
+                    // into the cache.
+                     None => on_change((key, value), CacheChange::Insertion),
                 }
             }
         }
 
         if !self.reset_on_next_modification {
-            extend_inner(&mut self.values, iter, on_change);
+            update_inner(&mut self.values, iter, on_change);
         } else {
-            let to_insert = iter.collect::<HashSet<T>>();
-            extend_inner(&mut self.values, to_insert.iter().map(|value| *value), on_change);
-            self.retain(&to_insert, on_change);
+            let to_insert = iter.collect::<IndexMap<K, V>>();
+            update_inner(
+                &mut self.values,
+                to_insert.iter().map(|(k, v)| (*k, v.clone())),
+                on_change,
+            );
+            self.update_intersection(&to_insert, on_change);
         }
         self.reset_on_next_modification = false;
     }
 
     pub fn remove<I, F>(&mut self, iter: I, on_change: &mut F)
-        where I: Iterator<Item = T>,
-              F: FnMut(T, CacheChange)
+    where
+        I: Iterator<Item = K>,
+        F: FnMut((K, V), CacheChange),
     {
         if !self.reset_on_next_modification {
-            for value in iter {
-                if self.values.remove(&value) {
-                    on_change(value, CacheChange::Removal);
+            for key in iter {
+                if let Some(value) = self.values.remove(&key) {
+                    on_change((key, value), CacheChange::Removal);
```

Comment:
In a follow-up, we should change this so that removals don't have values, by using a type like this:

```rust
enum CacheChange {
    Removal { key: K },
    Insertion { key: K, value: V },
    Modification { key: K, new_value: V },
};
```

Reply:
@briansmith yeah, I had considered making that change here but will do so in a follow-up instead.